File: 3_pin_sonar_sensor.cpp
//read the readme.txt file
#include <iostream>
#include <wiringPi.h>


using namespace std;
//attach sonar sensor's Vcc pin at 3.3V(recommended) or at 5V of RaspberryPi,  GND pin at RaspberryPi, signal pin at GPIO4(pin 7)
int main(void){
        wiringPiSetup();
        unsigned int timeout = 20000;
        int pin = 7;

        while(1){

		unsigned long int starttime, endtime;
                pinMode( pin, OUTPUT );
                digitalWrite( pin,0 );
                delayMicroseconds( 2 );
                digitalWrite( pin, 1 );
                delayMicroseconds( 5 );
                digitalWrite( pin, 0 );
                pinMode( pin, INPUT );
                bool goodread= true;
                unsigned long int watchtime= micros();
                while(digitalRead(pin)==0 && goodread){
                        starttime= micros();
                        if(starttime-watchtime > timeout){
                                goodread= false;
	                }
                }
                if(goodread){
                        watchtime=micros();
                        while(digitalRead(pin) && goodread){
                                endtime= micros();
                                if(endtime-watchtime>timeout){
                                        goodread= false;
                                }
                        }
                }

                if(goodread){
                        unsigned int duration= endtime - starttime;
                        float distance= (duration*34.0)/2000.0;
                        cout << distance << endl;
                }
        }
	return 0;
}

File: 3_pin_sonar_sensor.py
#read the readme.txt file
import time
import RPi.GPIO as GPIO

GPIO.setmode(GPIO.BOARD)
GPIO.setwarnings(False)
timeout = 0.020
#attach sonar sensor's Vcc pin at 3.3V(recommended) or at 5V of RaspberryPi,  GND pin at RaspberryPi, signal pin at GPIO4(pin 7) 
spin=7 # named pin7 as spin
while 1:
        GPIO.setup(spin, GPIO.OUT)
        #cleanup output
        GPIO.output(spin, 0)

        time.sleep(0.000002)

        #send signal
        GPIO.output(spin, 1)

        time.sleep(0.000005)  #change 5 microseconds delay to the timerequired by your sensor's trigger pin to send a pulse

        GPIO.output(spin, 0)

        GPIO.setup(spin, GPIO.IN)
        
        goodread=True
        watchtime=time.time()
        while GPIO.input(spin)==0 and goodread:
                starttime=time.time()
                if (starttime-watchtime > timeout):
                        goodread=False

        if goodread:
                watchtime=time.time()
                while GPIO.input(spin)==1 and goodread:
                        endtime=time.time()
                        if (endtime-watchtime > timeout):
                                goodread=False
        
        if goodread:
                duration=endtime-starttime
                distance=duration*34000/2
                print distance
